### DISCUSSION TITLE
REGEX-Support for "check:[]!

### DISCUSSION THREAD

**Discussion by T14D3** [Zrips/CMI#7267] [Enhancement]
I think it would be helpful if we had a way to use "check[]!" with REGEX-Expressions, for example when filtering if a string contains the name of the executing player and similar - since REGEX is pretty cool stuff I think the possibilities of this would be quite literally endless

**Comment by mrfloris** [comment]
I rathere have a more safe and less prone to breaking backend regex that is it's own specialized command that lets us get a true/false value and when it's true, to use a %result% in the command. something like this:

```
isin:$1==notch! asConsole! cmi usermeta add tempName %result%
check:%cmi_user_meta_tempName==bobby! msg! [playerName] we found the one and only bobby
check:%cmi_user_meta_tempName==notch! msg! [playerName] we found notch, lets kick him from the server
```
I don't know how complex people's inputs are that this is even needed, 
each parameter is separated by a space, can you give an example where you can't use check anymore?

/fruit banana apple meow
$0 = fruit
$1 = banana
$2 = apple
$3 = meow
and $1- = banana apple meow

check:$3==meow! msg! [playerName] success.

**Comment by MrCrazys** [comment]
> I think it would be helpful if we had a way to use "check[]!" with REGEX-Expressions, for example when filtering if a string contains the name of the executing player and similar - since REGEX is pretty cool stuff I think the possibilities of this would be quite literally endless

While this would likely be for more advanced server owners wanting to do more with specialized commands, I think this would be a useful thing to have, but still keep the current method available. There should be a way to define whether you want to use RegEx, perhaps. Of course, that's just my opinion.

**Comment by mrfloris** [comment]
More isn't always better.

I feel that giving players who already struggle with the syntax and very dynamic features of CMI very powerful expression options will lead to more damage than one might foresee. Regex isn't just a fun little solution to things. While the idea sounds great, I feel that it is more apt to keep it on the backend and give focused tools instead to help achieve 90% of the things that people want in edge cases.

One must keep in mind things such as User input validation, Faulty Regexes, Request Forgery, redirects and/or filter bypasses.. injection, players renaming items and triggering injections, .. It might be smarter to be strict, not publish expressions, validate input by having it on the backend, etc.

(imaginary -but perhaps unlikely- example)
I wouldn't want someone to rename an animal to `/../../../../../systemfilewhatever` and using a server owner made /ride command that checks if it's a pet with the same name as the owner, and then ending up the regex spitting out the result of the systemfilewhatever. Just because the server owner's /ride alias command doesn't go through the steps.

Or something like [] being used by CMI, but the expression is `[\w-]*$` to help with cname,. this would possibly allow the player to escape the [ `]\]+?(?=\.)`and change the expected filename, like databaseInfo.yml, .. nonsense like that.

**Reply by MrCrazys** [reply]
> More isn't always better.
> 
> I feel that giving players who already struggle with the syntax and very dynamic features of CMI very powerful expression options will lead to more damage than one might foresee. Regex isn't just a fun little solution to things. While the idea sounds great, I feel that it is more apt to keep it on the backend and give focused tools instead to help achieve 90% of the things that people want in edge cases.
> 
> One must keep in mind things such as User input validation, Faulty Regexes, Request Forgery, redirects and/or filter bypasses.. injection, players renaming items and triggering injections, .. It might be smarter to be strict, not publish expressions, validate input by having it on the backend, etc.
> 
> (imaginary -but perhaps unlikely- example) I wouldn't want someone to rename an animal to `/../../../../../systemfilewhatever` and using a server owner made /ride command that checks if it's a pet with the same name as the owner, and then ending up the regex spitting out the result of the systemfilewhatever. Just because the server owner's /ride alias command doesn't go through the steps.

I see what you mean. To be honest, I've never felt a need for it. I've been fine with how it is.

**Reply by mrfloris** [reply]
> > More isn't always better.
> > I feel that giving players who already struggle with the syntax and very dynamic features of CMI very powerful expression options will lead to more damage than one might foresee. Regex isn't just a fun little solution to things. While the idea sounds great, I feel that it is more apt to keep it on the backend and give focused tools instead to help achieve 90% of the things that people want in edge cases.
> > One must keep in mind things such as User input validation, Faulty Regexes, Request Forgery, redirects and/or filter bypasses.. injection, players renaming items and triggering injections, .. It might be smarter to be strict, not publish expressions, validate input by having it on the backend, etc.
> > (imaginary -but perhaps unlikely- example) I wouldn't want someone to rename an animal to `/../../../../../systemfilewhatever` and using a server owner made /ride command that checks if it's a pet with the same name as the owner, and then ending up the regex spitting out the result of the systemfilewhatever. Just because the server owner's /ride alias command doesn't go through the steps.
> 
> I see what you mean. To be honest, I've never felt a need for it. I've been fine with how it is.

It's quite handy if you have a string like UNBREAKING_3, to be able to isin:$1==UNBREAKING! and /cmi itemlore 3 {#gold}Unbreaking and the same for pulling out the "3" and visualize it with III

<img width="498" alt="Screen-Shot 2022-10-15 at 08 37 24" src="https://user-images.githubusercontent.com/28841349/195973097-bc324a6e-85e1-4d59-a849-d959adb4307f.png">
